File: robo_app/lib/data/repositories/watchlist_repository_impl.dart
import '../../domain/repositories/watchlist_repository.dart';
import '../datasources/remote/robo_remote_data_source.dart';

class WatchlistRepositoryImpl implements WatchlistRepository {
  WatchlistRepositoryImpl({required RoboRemoteDataSource remoteDataSource})
    : _remoteDataSource = remoteDataSource;

  final RoboRemoteDataSource _remoteDataSource;

  @override
  Future<void> addToWatchlist(String ticker) {
    return _remoteDataSource.addToWatchlist(ticker);
  }

  @override
  Future<List<String>> fetchWatchlist() {
    return _remoteDataSource.fetchWatchlist();
  }

  @override
  Future<void> removeFromWatchlist(String ticker) {
    return _remoteDataSource.removeFromWatchlist(ticker);
  }
}

File: Robo-back/src/infrastructure/repositories/watchlist_repository_impl.py
from __future__ import annotations

from sqlalchemy.orm import Session

from src.core.database import Watchlist
from src.domain.repositories.watchlist_repository import WatchlistRepository


class WatchlistRepositoryImpl(WatchlistRepository):
    def __init__(self, db: Session):
        self._db = db

    def get_all(self) -> list[str]:
        return [r.ticker for r in self._db.query(Watchlist).all()]

    def add(self, ticker: str) -> bool:
        exists = self._db.query(Watchlist).filter(Watchlist.ticker == ticker).first()
        if exists:
            return False
        self._db.add(Watchlist(ticker=ticker))
        self._db.commit()
        return True

    def remove(self, ticker: str) -> None:
        item = self._db.query(Watchlist).filter(Watchlist.ticker == ticker).first()
        if item:
            self._db.delete(item)
            self._db.commit()
